### PR TITLE
Database: Update SQLAlchemy to 1.4

### PR DESCRIPTION
**Pull request by bziemons**
- Replace query_bunches with Result.partitions
- Query the table instead of a column when deleting/updating
- Add synchronize_session to nested update queries
- Add try_drop_constraint for oracle
- Fix deprecation warning from empty 'or'
- Fix delete_qos_policy
- Stop leaking database objects from core
- Fix redirect_auth_oidc

Depends on #4688

### CODE REVIEW

**Review by rcarpa — COMMENTED** [review]
LGTM, but it mostly touches to code I've never seen, so I'm not very confident.

File: lib/rucio/api/subscription.py [hunk 1]
```diff
@@ -35,6 +35,9 @@
 from rucio.core import subscription
 
 
+SubscriptionRuleState = namedtuple('SubscriptionRuleState', ['account', 'name', 'state', 'count'])
```
Comment (mlassnig) [inline lib/rucio/api/subscription.py:38]: Is this only used in here? I thought we had something like this in common/ ?

File: lib/rucio/common/utils.py [hunk 2]
```diff
@@ -128,28 +127,6 @@ def build_url(url, path=None, params=None, doseq=False):
     return complete_url
 
 
-def oidc_identity_string(sub, iss):
```
Comment (mlassnig) [inline lib/rucio/common/utils.py:131]: why is this removed? (the other one below i understand)

File: lib/rucio/db/sqla/types.py [hunk 3]
```diff
@@ -46,8 +45,11 @@ class GUID(TypeDecorator):
     otherwise uses CHAR(32), storing as stringified hex values.
 
     """
+
     impl = CHAR
 
+    cache_ok = True
```
Comment (mlassnig) [inline lib/rucio/db/sqla/types.py:51]: what and where is this used for?

File: lib/rucio/db/sqla/util.py [hunk 4]
```diff
@@ -302,3 +302,17 @@ def json_implemented(session=None):
...
+def try_drop_constraint(constraint_name, table_name):
+    """
+    Tries to drop the given constrained and returns successfully if the
+    constraint already existed on Oracle databases.
+
+    :param constraint_name: the constraint's name
+    :param table_name: the table name where the constraint resides
+    """
+    try:
+        op.drop_constraint(constraint_name, table_name)
```
Comment (mlassnig) [inline lib/rucio/db/sqla/util.py:316]: does this work without explicitly specifying the constraint type? previously this was required to add

File: lib/rucio/common/utils.py (hunk 2 above)
Comment (bziemons) [reply lib/rucio/common/utils.py:131]: moved it to the oidc module, I dont think it belongs in client/server common code

File: lib/rucio/db/sqla/types.py (hunk 3 above)
Comment (bziemons) [reply lib/rucio/db/sqla/types.py:51]: That's a new SQLAlchemy feature: https://docs.sqlalchemy.org/en/14/core/custom_types.html#sqlalchemy.types.TypeDecorator.cache_ok

(not specifying it spammed warnings)

File: lib/rucio/db/sqla/util.py (hunk 4 above)
Comment (bziemons) [reply lib/rucio/db/sqla/util.py:316]: that is only necessary for mysql, according to the documentation https://alembic.sqlalchemy.org/en/latest/ops.html?highlight=drop_constraint#alembic.operations.Operations.drop_constraint (and confirmed by the tests)

File: lib/rucio/api/subscription.py (hunk 1 above)
Comment (bziemons) [reply lib/rucio/api/subscription.py:38]: can you point me to it? I had to replace the KeyedTuple, because that is no more

Comment (bziemons) [reply lib/rucio/api/subscription.py:38]: can you point me to it? I had to replace the KeyedTuple, because that is no more

Comment (mlassnig) [reply lib/rucio/api/subscription.py:38]: https://github.com/rucio/rucio/blob/master/lib/rucio/common/constants.py

--> e.g. it has the FTS States there

Comment (bziemons) [reply lib/rucio/api/subscription.py:38]: yeah, this is only used here and does not break the API, so that you can still use

```py3
for row in list_subscription_rule_states():
    print(row.account, row.name, row.state, row.count)
```
or similar code. Contrary to `FTS_STATES`, this isn't constant so I would not put it into the constants module

**Review by mlassnig — APPROVED** [review]

**Review by bari12 — APPROVED** [review]
👍 I am a bit skeptical about the relative release specifier in the requirements. I agree that's what should be used, but I think we did have issues in the past even with patch release changes.
